src/Razor/src/Microsoft.CodeAnalysis.Remote.Razor/DocumentMapping/RemoteDocumentMappingService.cs

```diff
@@ -46,31 +47,18 @@ internal sealed class RemoteDocumentMappingService(
             return (generatedDocumentUri, generatedDocumentRange);
         }
 
-        var solution = originSnapshot.TextDocument.Project.Solution;
-        if (!solution.TryGetRazorDocument(razorDocumentUri, out var razorDocument))
-        {
-            return (generatedDocumentUri, generatedDocumentRange);
-        }
-
-        var razorDocumentSnapshot = _snapshotManager.GetSnapshot(razorDocument);
-
-        var razorCodeDocument = await razorDocumentSnapshot
-            .GetGeneratedOutputAsync(cancellationToken)
-            .ConfigureAwait(false);
-
-        if (razorCodeDocument is null)
-        {
-            return (generatedDocumentUri, generatedDocumentRange);
-        }
-
-        if (!razorCodeDocument.TryGetGeneratedDocument(generatedDocumentUri, _filePathService, out var generatedDocument))
-        {
-            return Assumed.Unreachable<(Uri, LinePositionSpan)>();
-        }
-
-        if (TryMapToHostDocumentRange(generatedDocument, generatedDocumentRange, MappingBehavior.Strict, out var mappedRange))
+        var project = originSnapshot.TextDocument.Project;
+        var solution = project.Solution;
+        if (await project.TryGetHintNameFromGeneratedDocumentUriAsync(generatedDocumentUri, cancellationToken).ConfigureAwait(false) is { } hintName &&
+            await project.TryGetSourceGeneratedDocumentFromHintNameAsync(hintName, cancellationToken).ConfigureAwait(false) is { } generatedDocument &&
+            await _snapshotManager.GetSnapshot(generatedDocument.Project).TryGetCodeDocumentFromGeneratedHintNameAsync(hintName, cancellationToken).ConfigureAwait(false) is { } razorCodeDocument &&
+            TryMapToHostDocumentRange(razorCodeDocument.GetCSharpDocument(), generatedDocumentRange, MappingBehavior.Strict, out var mappedRange))
```

Comment:
Consider breaking this into separate statements to allow breakpoints to be set. So much hopping around async state machines within the condition of an if statement is a bit... extreme. 😆 If you really want to keep them like this, please add some sensible line breaks since this is what I'm reviewing:

![image](https://github.com/user-attachments/assets/af667607-b47f-4d03-9557-1f436e1d8a4d)